serializer_impl.hh

```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
+            return _idx - it._idx;
+        }
+    };
+
+    // Forward iterator
+    class iterator {
+        utils::input_stream _in;
+        uint32_t _idx = 0;
+        std::optional<T> _item;
+
+        iterator(const utils::input_stream& in, uint32_t idx, std::optional<T>&& item)
+            : _in(in)
+            , _idx(idx)
+            , _item(std::move(item))
+        { }
+
+        friend class range_deserializer;
+    protected:
+        iterator(const utils::input_stream& in)
+            : _in(in)
+            , _idx(0)
+        { }
+
+    private:
+        static size_type read_frame_size(utils::input_stream& in) {
+            auto sz = deserialize(in, boost::type<size_type>());
+            if (sz < sizeof(size_type)) {
+                throw std::runtime_error(fmt::format("IDL frame truncated: expected to have at least {} bytes, got {}", sizeof(size_type), sz));
+            }
+            return sz - sizeof(size_type);
+        }
+
+   public:
+        iterator() = default;
+
+        iterator(iterator&& o) noexcept
+            : _in(std::move(o._in))
+            , _idx(o._idx)
+            , _item(std::exchange(o._item, std::nullopt))
+        { }
+
+        iterator(const iterator& o) = default;
+
+        iterator& operator=(iterator&&) = default;
+        iterator& operator=(const iterator&) = default;
+
+        bool operator==(const sentinel& s) noexcept {
+            return _idx >= s._idx;
+        }
+
+        bool operator==(const iterator& it) noexcept {
+            return _idx == it._idx;
+        }
+
+        // Return const& as the input stream cannot be modified
+        const T& operator*() {
+            if (!_item) {
+                _item.emplace(deserialize(_in, boost::type<T>()));
+            }
+            return *_item;
+        }
+        // Return const* as the input stream cannot be modified
+        const T* operator->() {
+            return &this->operator*();
+        }
+
+        iterator& operator++() {
+            if (!_item) {
+                auto len = read_frame_size(_in);
+                _in.skip(len);
+            } else {
+                _item.reset();
+            }
+            ++_idx;
+            return *this;
+        }
+        iterator operator++(int) {
+            if (_item) {
+                auto pre = iterator(_in, _idx, std::exchange(_item, std::nullopt));
+                // If item was already read we don't need to skip in the input_stream.
+                ++_idx;
+                return pre;
+            } else {
+                auto pre = iterator(_in, _idx);
+                // Skip the current item in the input_stream.
+                ++*this;
+                return pre;
+            }
+        }
+
+        ssize_t operator-(const iterator& it) const noexcept {
+            return _idx - it._idx;
+        }
+    };
+
+    iterator begin() {
+        return iterator(_in);
+    }
+    sentinel end() {
+        return sentinel(_size);
+    }
+};
```

Comment:
Will this work also for `std::unordered_map<K, V>`, making `T=std::pair<K, V>`?

Reply:
Hmm, not as is,
We may be able to use the same template by defining the view as T=std::pair<K_view, V_view>
Maybe it'd be best to rename the class to `vector_deserializer` until iterative deserialization for maps is implemented.